### ISSUE TITLE
Add generic dropdown component 

### ISSUE THREAD

**Issue by iHiD** [exercism/website#400]
We have a generic tooltip component now, which we populate with HTML when it renders. I want to replicate this for a dropdown component.

This will be the component that is triggered when someone clicks on the top-right profile picture to show their main menu, or the notifications icon to see their notifications.

- This to be attached to an element, which should be clickable [1]
- Clicking it should reveal a visible div (its dropdown) [2]
- The dropdown should be anchored to at the right-hand-side [3], offset by an optional offset passed as a data-attribute (default zero if missing)
- The content of the div should either:
  - be pulled async (same as the modal, e.g. for notifications)
  - be prerendered as a data-attribute on the clickable element (e.g. for the profile menu) [2]
- Clicking on something else on the page should hide the dropdown
- Clicking on the original element should hide dropdown
- It should follow accessible guidelines (cc @SleeplessByte for any input you want to put into this before it's made 💙)

[1] 
<img width="74" alt="Screenshot 2021-01-05 at 23 19 18" src="https://user-images.githubusercontent.com/286476/103710542-28f4cb00-4fad-11eb-8340-398bcc4ce2b3.png">

[2] 
<img width="273" alt="Screenshot 2021-01-05 at 23 19 22" src="https://user-images.githubusercontent.com/286476/103710551-2eeaac00-4fad-11eb-9c3b-e2b5a296a4e0.png">

[3] 
<img width="471" alt="Screenshot 2021-01-05 at 23 26 30" src="https://user-images.githubusercontent.com/286476/103710642-74a77480-4fad-11eb-80f4-b8e579d7759e.png">

---

@ErikSchierboom This one's for you.

**Comment by SleeplessByte** [comment]
Ok. So. 

Attached ([aria-dropdown.zip](https://github.com/exercism/v3-website/files/5784128/aria-dropdown.zip)) is a ZIP file with LOADS of commentary on how everything works. It is an spec compliant implementation *without submenus*. It works by looking for `[data-overflow="menu"]` (not in that code), and instantiating `new MenuButton(component)`. It calls `.destroy` when it unmounts. The button ensures the menu is created, etc.

## Authoring Practices

This will follow the `menu` component (that's what this is called, canonically), found here: https://www.w3.org/TR/wai-aria-practices/#menu

> A menu is a widget that offers a list of choices to the user, such as a set of actions or functions. A menu is usually opened, or made visible, by activating a menu button, choosing an item in a menu that opens a sub menu, or by invoking a command, such as Shift + F10 in Windows, that opens a context specific menu. When a user activates a choice in a menu, the menu usually closes unless the choice opened a submenu.

## Required keyboard interaction

Definitions:

- menu button: button that opens/closes the menu (source of the menu, usually also the anchor of the menu)
- menu: all the menu items together, excluding the menubutton, defaults to vertical orientation
- menubar: a permanently visible menu (not the same as a tabbar	etc.), defaults to horizontal orientation			
- menuitem: an item in the menu, can be a link, a button.
- menuitemradio: an item in the menu, as a radio button
- menuitemcheckbox: an item in the menu, as a checkbox
- `[character]`*: Any key that corresponds to a printable character 

| key | description |
|---|---|
| Enter | Activates the item and closes the menu. |
| Space | Handling space is _optional_. I *recommend* delegating to `Enter` in this case. If you must know, the list is below this table. |
| Down Arrow | INSIDE menu: Move focus to next item. Recommended: Wrap to the first one. ON menubutton: open menu and places focus on the first item |
| Up Arrow | INSIDE menu: Move focus to previous item. Recommended: Wrap to the last one. ON menubutton: open menu and places focus on the last item |
| Home | Required when DOWN/UP don't wrap; Recommended: Moves focus to the first item in the current menu |
| End | Required when DOWN/UP don't wrap; Recommended: Moves focus to the last item in the current menu |
| `[character]`* | Recommended: Move focus to the next menu item in the current menu whose label begins with that printable character. |
| Escape | Close the menu that contains focus and return focus to the element or context, e.g., menu button or parent menuitem, from which the menu was opened. |
| Tab | Moves focus to the next element in the tab sequence, and if the item that had focus is not in a menubar, closes its menu and all open parent menu containers. |
| Shift + Tab | Moves focus to the previous element in the tab sequence, and if the item that had focus is not in a menubar, closes its menu and all open parent menu containers. |

- Right and left arrows only are required for submenus, which we don't have. 
- Enter behaves different when on a menuitem with a submenu: opens the submenu and places focus on its first item. 

### Handling Space

- (Optional): When focus is on a menuitemcheckbox, changes the state without closing the menu.
- (Optional): When focus is on a menuitemradio that is not checked, without closing the menu, checks the focused menuitemradio and unchecks any other checked menuitemradio element in the same group.
- (Optional): When focus is on a menuitem that has a submenu, opens the submenu and places focus on its first item.
- (Optional): When focus is on a menuitem that does not have a submenu, activates the menuitem and closes the menu.

### Notes about focus

- The entire component (menu button + menu) should be a _single tab stop_. 
- Disabled menu items **are** focusable but cannot be activated.
- A separator in a menu is not focusable or interactive.
- If a menu is opened or a menubar receives focus as a result of a context action, Escape or Enter may return focus to the invoking context. For example, a rich text editor may have a menubar that receives focus when a shortcut key, e.g., alt + F10, is pressed while editing. In this case, pressing Escape or activating a command from the menu may return focus to the editor.
- Left/Right and Up/Down arrows become Up/Down and Left/Right if menu is horizontal or menubar is vertical
- Returning focus to the item that open the menu is _very_ important for screenreaders

## Aria states / attributes

- [`role="menu"`](https://www.w3.org/TR/wai-aria-1.1/#menu) for the item that is the container for the menuitems
- [`role="menuitem"`](https://www.w3.org/TR/wai-aria-1.1/#menuitem) on each menuitem
- [`aria-haspopup="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-haspopup) (or "menu") on the menu button
- [`aria-expanded="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-expanded) on the menu button, when the menu is visible, false otherwise (recommendation is to remove the attribute completely when the menu is hidden).
- each menuitem has `tabindex=-1` (rules different for menubar, but okay), so that they don't become tab-stops
- [`aria-checked="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-checked) on each _active_/_checked_ menuitemradio and menuitemcheckbox
- [`aria-disabled="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-disabled) on each menuitem if its disabled
- [`role="separator"`](https://www.w3.org/TR/wai-aria-1.1/#separator) for separators
- An element with role menu either has:
  - `aria-labelledby` set to a value that refers to the menuitem or button that controls its display.
  - A label provided by `aria-label`.
- [`aria-controls="..."`](https://www.w3.org/TR/wai-aria-1.1/#aria-controls) on the menu button. This is REQUIRED if the menu button is HTML structurally NOT a sibling of the menu itself, so my recommendation is to always add it.

## Orientations

As mentioned before, orientation different than the default changes the way the keyboard interaction works, but also requires `aria-orientation="..."` to be set. It doesn't need to be set for the default orientations.

## Element that _triggers_ the dropdown (menu button)

This will follow the `menu button` component, found here: 
https://www.w3.org/TR/wai-aria-practices/#menubutton

Prefer a `button`, **never* an empty link. You may use a link (`a`), if that link, when JavaScript is disabled, refreshes the page with the new state (e.g. opens/closes the menu).

The _most_ semantic way to tackle it would be to have whatever you want to display be semantically valid, and "upgrade" to a button if JavaScript is available.

The following is mentioned already in the keyboard section, but to re-iterate:

With focus on the button:
- Enter: opens the menu and places focus on the first menu item.
- Space: Opens the menu and places focus on the first menu item.
- (Optional) Down Arrow: opens the menu and moves focus to the first menu item.
- (Optional) Up Arrow: opens the menu and moves focus to the last menu item.

## Toggle on hover?

TL;DR: no.

The longer TL;DR version is: sure, but only if you handle every case correctly and THAT is really hard to do. It's better to NOT do it.

From [A11Y Developer Guide](https://www.accessibility-developer-guide.com/examples/widgets/dropdown/#toggle-on-hover) which I 100% agree with:

> There is a big debate whether dropdown menus should open on hover. First of all: unintentional hovering may cause flickering of elements. But there are also confusing (if not disrupting or even misleading) behaviours attached to dropdowns that open (or close) on hover.
>
> Look at this scenario: a user hovers over a dropdown and clicks on it, but nothing seems to happen. This is due to the fact that the dropdown already was triggered to open by hover, and clicking on it closes it immediately again.
>
> An even worse scenario is this: a user wants to click an element right below a closed dropdown. While moving the mouse over the dropdown (which is of no interest to the user at this time), the hover event opens the dropdown and overlaps the target element, causing the fast user to click on an element in the dropdown instead.
> 
> The worst are elements that toggle a dropdown on hover and navigate to a website by activation. This is extremely confusing for all sorts of users (be it with or without special needs), and pretty hard to implement in an accessible way.
> 
> There are a lot of tricks and workarounds to minimise the problems depicted above, like using timeouts or visual effects. But in general, the problems remain, and the workarounds potentially lead to new side effects (or even less intuitive behaviours).
> 
> In our opinion, hovering events should only be used for purely visual purposes, and toggling the visibility of content is by no means purely visual. So it is much better to only open (and close) dropdowns on dedicated activation by the user.

## Specific implementation

For the menu in the screenshot, I recommend making the "header" a selectable link. If you _don't_ want that, you can consider making it a "decorative" element, with the "open external" button being a link that is _HTML structurally_ at the *bottom* of the list, not at the top. 

In other words:

- 4 links (header, my journey, profile, settings)
- 1 button (sign out)
